NorthstarDedicatedTest/masterserver.cpp

```diff
@@ -545,6 +567,134 @@ void MasterServerManager::AuthenticateWithServer(char* uid, char* playerToken, c
 	requestThread.detach();
 }
 
+std::string MasterServerManager::GetServerPing(char* uid, char* playerToken, RemoteServerInfo* server)
+{
+	std::thread requestThread([this, uid, playerToken, server]()
+		{
+			httplib::Client http(Cvar_ns_masterserver_hostname->m_pszString);
+			http.set_connection_timeout(25);
+
+			spdlog::info("Attempting authentication with server of id \"{}\"", server->id);
+
+			if (auto result = http.Post(fmt::format("/client/auth_with_server?id={}&playerToken={}&server={}&password=", uid, playerToken, server->id).c_str()))
+			{
+
+				rapidjson::Document connectionInfoJson;
+				connectionInfoJson.Parse(result->body.c_str());
+
+				if (connectionInfoJson.HasParseError())
+				{
+					spdlog::error("Failed reading masterserver authentication response: encountered parse error \"{}\"", rapidjson::GetParseError_En(connectionInfoJson.GetParseError()));
+					goto REQUEST_END_CLEANUP;
+				}
+
+				if (!connectionInfoJson.IsObject())
+				{
+					spdlog::error("Failed reading masterserver authentication response: root object is not an object");
+					goto REQUEST_END_CLEANUP;
+				}
+
+				if (connectionInfoJson.HasMember("error"))
+				{
+					spdlog::error("Failed reading masterserver response: got fastify error response");
+					spdlog::error(result->body);
+					goto REQUEST_END_CLEANUP;
+				}
+
+				if (!connectionInfoJson["success"].IsTrue())
+				{
+					spdlog::error("Authentication with masterserver failed: \"success\" is not true");
+					goto REQUEST_END_CLEANUP;
+				}
+
+				if (!connectionInfoJson.HasMember("success") || !connectionInfoJson.HasMember("ip") || !connectionInfoJson["ip"].IsString() || !connectionInfoJson.HasMember("port") || !connectionInfoJson["port"].IsNumber() || !connectionInfoJson.HasMember("authToken") || !connectionInfoJson["authToken"].IsString())
+				{
+					spdlog::error("Failed reading masterserver authentication response: malformed json object");
+					goto REQUEST_END_CLEANUP;
+				}
+
+				spdlog::info(fmt::format("Server with ID {} has address {}:{}", server->id, connectionInfoJson["ip"].GetString(), connectionInfoJson["port"].GetInt()).c_str());
+				std::string ping = MasterServerManager::SendPing(connectionInfoJson["ip"].GetString(), server);
+				return ping;
+			}
+			else
+			{
+				spdlog::error("Failed authenticating with server: error {}", result.error());
+				return (std::string)("NaN");
+			}
+
+			REQUEST_END_CLEANUP:
+			spdlog::error("Finished authenticating with server");
+			return (std::string)("NaN");
+
+		});
+
+	requestThread.detach();
+
+	return (std::string)("NaN");
+}
+
+std::string MasterServerManager::SendPing(const char* ip, RemoteServerInfo* server)
+{
+	// Declare and initialize variables
+
+	HANDLE hIcmpFile;
+	unsigned long ipaddr = INADDR_NONE;
+	DWORD dwRetVal = 0;
+	char SendData[32] = "Data Buffer";
+	LPVOID ReplyBuffer = NULL;
+	DWORD ReplySize = 0;
+
+	ipaddr = inet_addr(ip);
```

Comment:
Maybe use InetPton instead here also.

Reply:
Any idea how I would go about changing it to use InetPton? This is basically my first time using C++ so I'm clueless